src/crimson/admin/admin_socket.cc

```diff
@@ -0,0 +1,696 @@
+}
+
+seastar::future<> AdminSocket::init(const std::string& path)
+{
+  if (!path.length()) {
+    logger().error(
+      "{}: Admin Socket socket path missing from the configuration", __func__);
+    return seastar::now();
+  }
+
+  logger().debug("{}: asok socket path={}", __func__, path);
+  auto sock_path = seastar::socket_address{ seastar::unix_domain_addr{ path } };
+
+  std::ignore = internal_hooks().then([this,
+                                       sock_path](AsokRegistrationRes reg_res) {
+    return seastar::do_with(
+      seastar::engine().listen(sock_path),
+      [this](seastar::server_socket& lstn) {
+        m_server_sock = &lstn;  // used for 'abort_accept()'
+        return do_until_gate(
+                 arrivals_gate,
+                 [&lstn, this]() {
+                   return lstn.accept().then(
+                     [this](seastar::accept_result from_accept) {
+                       seastar::connected_socket cn =
+                         std::move(from_accept.connection);
+
+                       return do_with(
+                         cn.input(), cn.output(), std::move(cn),
+                         [this](auto& inp, auto& out, auto& cn) {
+                           return handle_client(inp, out).finally([] {
+                             ;  // left for debugging
+                           });
+                         });
+                     });
+                 })
+          .then([] {
+            logger().debug("AdminSocket::init(): admin-sock thread terminated");
+            return seastar::now();
+          });
+      });
+  });
+
+  return seastar::make_ready_future<>();
+}
+
+seastar::future<> AdminSocket::stop()
+{
+  // RF Dec-2019: abort_accept(), at least when applied to a UNIX-domain socket
+  // with an 'epoll' backend, will not correctly abort a Seastar accept():
+  // epoll_wait() will signal EPOLLRDHUP+EPOLLIN, and that would be handled
+  // incorrectly.
+
+  if (m_server_sock && !arrivals_gate.is_closed()) {
+    // note that we check 'is_closed()' as if already closed - the server-sock
+    // may have already been discarded
+    m_server_sock->abort_accept();
+    m_server_sock = nullptr;
+  }
+
+  return seastar::futurize_apply([this] { return arrivals_gate.close(); })
+    .then_wrapped([this](seastar::future<> res) {
+      if (res.failed()) {
+        std::ignore = res.handle_exception([](std::exception_ptr eptr) {
+          return seastar::make_ready_future<>();
+        });
+      }
+      return seastar::make_ready_future<>();
+    })
+    .handle_exception(
+      [](std::exception_ptr eptr) { return seastar::make_ready_future<>(); })
+    .finally([]() { return seastar::make_ready_future<>(); });
+}
+
+// ///////////////////////////////////////
+// the internal hooks
+// ///////////////////////////////////////
+
+/**
+  The response to the '0' command is not formatted, neither as JSON or
+  otherwise.
+*/
+class The0Hook : public AdminSocketHook {
```

Comment:
i don't think we need to support the "0" command. it's not sent by "ceph" cli anymore. the only caller which sends "0" is "AdminSocket.SendNoOp" test in `src/test/admin_socket.cc`. i think we can remove that test in a separated commit.

Reply:
I'd like to remove the support for "0" in a separate PR. There is an efficiency price to pay for delaying this specific PR.